User-feedback functionality

PascalLike opened the pull request:
Starting with https://github.com/geonetwork/core-geonetwork/wiki/Userfeedback-proposal
we created a functionality for user feedback on metadata records.

The changes are implemented starting by 4 angular directives:

1. **gnUserfeedbacklasthome** - Shows messages and ratings on homepage
2. **gnUserfeedback** - Shows messages and ratings on metadata view
3. **gnUserfeedbackfull** - See details about messages and ratings associated with a metadata record and manages publication of feedbacks from anonymous users.
4. **gnUserfeedbacknew** - To insert a comment and a rating.

The creation of user feedbacks is allowed for every user. For feedbacks from not signed in users a basic workflow will be applied (the comment will be hidden to non-reviewers until the reviewer approval).

![list](https://user-images.githubusercontent.com/1323093/32002805-a102fe18-b99d-11e7-9154-74ccbfcb8413.png)

**gnUserfeedbacklasthome** is added on the home page.

![home](https://user-images.githubusercontent.com/1323093/32002817-a8eb157a-b99d-11e7-9632-16ec755ecadd.png)

**gnUserfeedback**, **gnUserfeedbackfull** and **gnUserfeedbacknew** are integrated in the record view.

![new](https://user-images.githubusercontent.com/1323093/32002828-b0327cba-b99d-11e7-9239-5fd019a4b83c.png)

The GN API now contains new methods for UF (api/{version/}userfeedback/...): Creating, deleting, publishing and listing. Some methods returns statistic used by the different views implemented in the directives.
![api](https://user-images.githubusercontent.com/1323093/32002711-68cb3614-b99d-11e7-8da5-78ced5969cc0.png)
The feedbacks are stored in new tables in gn's database.

The functionality is disabled by default but can be activated in settings page. Is very similar to the rating functionality, so the setting for local rating now supports these 3 options:
**off** - no rating
**basic** - the current 5 star rating
**advanced** - the new user feedback functionality

![settings](https://user-images.githubusercontent.com/1323093/32002737-75712130-b99d-11e7-9971-efa005a3b5e5.png)

There is another function called user feedback and it's a bit ambiguos. The official documentation should be updated.

This software has been improved under the [ELISE action](https://ec.europa.eu/isa2/actions/elise_en), which is supported by the ISA² Programme. ISA² is a EUR 131 million programme of the European Commission, supporting the modernisation of public administrations in Europe through the development of eGovernment solutions. More than 20 solutions are already available, with more to come soon. All solutions are open source and available free of charge to any interested public administration in Europe.

Delawen commented:
Moving this to 3.4.2 because I want to close today 3.4.1 to provide a stable version without some relevant bugs and this is a big change to be sure....

josegar74 commented on core/src/main/java/org/fao/geonet/kernel/DataManager.java:
```diff
@@ -2226,6 +2267,27 @@ private void deleteMetadataFromDB(ServiceContext context, String id) throws Exce
+
+        Metadata metadata = metadataRepository.findOne(metadataId);
+        
+        List<UserFeedback> list = userfeedbackRepository.findByMetadata_Uuid(metadata.getUuid());
```
Can't be added a method `deleteByMetadataUuid` to `UserFeedbackRepository`?

josegar74 commented on services/src/main/java/org/fao/geonet/api/userfeedback/service/UserFeedbackDatabaseService.java:
```diff
@@ -0,0 +1,218 @@
+    final UserFeedbackRepository userFeedbackRepository = ApplicationContextHolder.get()
+        .getBean(UserFeedbackRepository.class);
+
+    // TODO: Manage UserFeedback not found
```
Please check this.

josegar74 commented on web/src/main/webapp/WEB-INF/classes/setup/sql/migrate/v340/migrate-default.sql:
```diff
@@ -5,5 +5,7 @@ INSERT INTO Settings (name, value, datatype, position, internal) VALUES
 
 INSERT INTO Settings (name, value, datatype, position, internal) VALUES ('system/metadata/validation/removeSchemaLocation', 'false', 2, 9170, 'n');
 
+UPDATE Settings SET name = 'system/localrating/enable', datatype = 0 WHERE id = 211;
```
In the UI the values seem `off`, `basic`, `advanced`, I think the migration script should handle to change the previous values `false`, `true`.

Also should be added to the proper migration file, 340 is no longer valid when the PR is merged.

PascalLike commented:
@josegar74 I updated according to your review. If it's fine for you, please, approve. 

I just updated with last code from 3.4.x. There's a lot of code impacted for this new feature and it's expensive to keep it updated.

fxprunayre commented:
@PascalLike do you think we should have this in 3.4.2 ? or make the release first, then merge ?

fxprunayre commented on services/src/main/java/org/fao/geonet/api/userfeedback/UserFeedbackAPI.java:
```diff
@@ -0,0 +1,424 @@
+     * @throws Exception the exception
+     */
+    @ApiOperation(value = "Provides an average rating for a metadata record", nickname = "getMetadataUserComments")
+    @RequestMapping(value = "/metadata/{uuid}/userfeedbackrating", produces = MediaType.APPLICATION_JSON_VALUE, method = RequestMethod.GET)
```
I would renamed /metadata by /records to be consistent with the API

fxprunayre commented on services/src/main/java/org/fao/geonet/api/userfeedback/UserFeedbackAPI.java:
```diff
@@ -0,0 +1,424 @@
+    @ResponseStatus(value = HttpStatus.OK)
+    @ResponseBody
+    public RatingAverage getMetadataRating(@PathVariable(value = "uuid")
+    final String metadataUuid, final HttpServletRequest request, final HttpServletResponse response, final HttpSession httpSession)
```
Add ```@ApiIgnore``` annotation to avoid non API parameters to be displayed in the doc 

```
@ApiIgnore final HttpServletRequest request, @ApiIgnore final HttpServletResponse response, @ApiIgnore final HttpSession httpSession
```

![image](https://user-images.githubusercontent.com/1701393/38668443-40c06820-3e44-11e8-92b7-538abd6ae0b6.png)

fxprunayre commented on services/src/main/java/org/fao/geonet/api/userfeedback/UserFeedbackAPI.java:
```diff
@@ -0,0 +1,424 @@
+
+            final IUserFeedbackService userFeedbackService = getUserFeedbackService();
+
+            final String uuid = request.getParameter("target");
```
```target``` and ```maxnumber``` parameters are not part of the method signature and they are not displayed in the service doc 
![image](https://user-images.githubusercontent.com/1701393/38668930-5b795892-3e45-11e8-9186-c3941c1a8da8.png)

Use ```@RequestParam``` annotation ?

Also "target" may be renamed to "record" maybe as it is a metadata record.

At some point we may need to add paging here - but this is not part of this PR - we should maybe plan for adding this for users/groups/... API.

fxprunayre replied:
If target is mandatory, should we move that service to /records/{uuid}/userfeedback ? would be more consistent with userfeedbackrating

fxprunayre replied:
target is not mandatory, so maybe not. or should we have both ?

PascalLike replied:
None of them is mandatory. I added them (uuid, previously target, and maxnumber) as not requested parameters

PascalLike commented:
I would really like to see it merged in 3.4.2, so if it's fine for you please go ahead!

fxprunayre commented on web-ui/src/main/resources/catalog/components/userfeedback/GnUserfeedbackDirective.js:
```diff
@@ -0,0 +1,411 @@
+
+                $http({
+                  method: 'GET',
+                  url: '../api/userfeedback?full=true&uuid=' + scope.metatdataUUID,
```
I don't see this full parameter in the server side. What is it for ?

PascalLike replied:
No, it's not used anymore

PascalLike commented:
Thanks for the review (and the commit). Please let me know if it's ok now

fxprunayre commented on web-ui/src/main/resources/catalog/components/userfeedback/GnUserfeedbackDirective.js:
```diff
@@ -0,0 +1,411 @@
+                  console.log('DELETED ' + id);
+
+                  $http.delete('../api/userfeedback/' + id).success(function(data, status) {
+                    console.log(data);
```
Do we need all those console.log ?

fxprunayre commented on web-ui/src/main/resources/catalog/components/userfeedback/GnUserfeedbackDirective.js:
```diff
@@ -0,0 +1,411 @@
+
+              };
+
+              scope, showPopover = function(info) {
```
Looks odd and useless no ?

fxprunayre commented on web-ui/src/main/resources/catalog/components/userfeedback/GnUserfeedbackDirective.js:
```diff
@@ -0,0 +1,411 @@
+
+              };
+
+              function validateEmail(email) {
```
This function is never used no ? Email check is done in line 345 ?

fxprunayre commented on web-ui/src/main/resources/catalog/components/userfeedback/GnUserfeedbackDirective.js:
```diff
@@ -0,0 +1,411 @@
+
+                $http.post('../api/userfeedback', data).success(function(data, status) {
+                  console.log(data);
+                  $window.location.reload();
```
When user post the feedback, the page is fully reload - we don't need that. Only refresh the list of feedback no ?

fxprunayre commented on web-ui/src/main/resources/catalog/components/userfeedback/partials/userfeedbacknew.html:
```diff
@@ -0,0 +1,215 @@
+          <input type="hidden" data-ng-model="uf.ratingAVG">
+          <button type="button" class="btn btn-default" data-dismiss="modal"
+            data-translate="">GUFclose</button>
+          <button type="button" class="btn btn-primary"
```
Would be good to add recaptcha to avoid automatic submission of the form ?

fxprunayre commented on web-ui/src/main/resources/catalog/views/default/templates/home.html:
```diff
@@ -172,6 +172,12 @@ <h4>
           </form>
         </tab>
+       
+        <tab heading="Comments" active="infoTabs.commentsalt.active" data-ng-show="isUserFeedbackEnabled">
```
Not translated

fxprunayre commented on web-ui/src/main/resources/catalog/components/userfeedback/partials/userfeedbacklasthome.html:
```diff
@@ -0,0 +1,32 @@
+          - / - <span class="fa fa-fw fa-star">&nbsp;</span>
+        </div>
+        <div class="gn-userfeedback-reply">
+          <a>{{row.authorName}}</a><br>on {{row.date | date:'medium'}}
```
Link on user is not set - should we remove ?

At some point, we may group new messages on a same record ?
![image](https://user-images.githubusercontent.com/1701393/38680214-04fcd8d8-3e66-11e8-87a8-cf1e7861911a.png)

fxprunayre commented on web-ui/src/main/resources/catalog/locales/en-search.json:
```diff
@@ -419,5 +419,52 @@
+  "GUFsave": "Save",
+  "GUFfeedbackOn": "Feedback on:",
+  "GUFnoComments": "No comments",
+  "GUFaddRating": "Review this dataset",
```
Maybe we should make this more generic "Review this resource" because you may have a service, map, ...
![image](https://user-images.githubusercontent.com/1701393/38681980-4bf1c57e-3e6a-11e8-9143-7d56c67793c6.png)

or adapt the label depending on the hierarchylevel ?

fxprunayre commented on web-ui/src/main/resources/catalog/locales/en-search.json:
```diff
@@ -419,5 +419,52 @@
+  "GUFfeedbackOn": "Feedback on:",
+  "GUFnoComments": "No comments",
+  "GUFaddRating": "Review this dataset",
+  "GUFplusAddRating": "<span class=\"fa fa-plus\"></span> Review this dataset",
```
We usually put the HTML in templates not in translation to keep translation as easy as possible

PascalLike commented:
Done. I really like this last improvements, now it looks better. Thank you

fxprunayre commented:
@PascalLike , We also need to index this depending on the settings in here https://github.com/geonetwork/core-geonetwork/blob/3.4.x/core/src/main/java/org/fao/geonet/kernel/DataManager.java#L683. Currently only the basic rating is displayed whatever the config. So this is misleading and the stars in search results does not represent the average of rating.


eg. no start displayed

![image](https://user-images.githubusercontent.com/1701393/39116471-d441e652-46e4-11e8-9494-c226f70fe6d0.png)

on a record with one user feedback

![image](https://user-images.githubusercontent.com/1701393/39116478-d833378e-46e4-11e8-8c9d-0cc9b3a2c791.png)


Do you have time to cover that ?

fxprunayre commented on services/src/main/java/org/fao/geonet/api/userfeedback/UserFeedbackUtils.java:
```diff
@@ -0,0 +1,596 @@
+        String metadataTitle;
+        try {
+            final Element element = input.getMetadata().getXmlData(false);
+            final XPath xpath = XPath.newInstance("gmd:identificationInfo/*/gmd:citation/gmd:CI_Citation/gmd:title/gco:CharacterString");
```
This will not work for records not in ISO19139.

fxprunayre commented on domain/src/main/java/org/fao/geonet/domain/userfeedback/Citation.java:
```diff
@@ -0,0 +1,119 @@
+
+    /** The date. */
+    @Column
+    private Date date;
```
Would have been better to use same encoding for date as other classes ie. ISODate
